File: www/js/app/services/Online.js
(function () {
  'use strict';

  angular
    .module('AngularChatApp')
    .factory('Online', ['$http', '$rootScope', function($http, $rootScope){
            return {
                        setOnline: setOnline,
                        setOffline: setOffline,
                        addToOnline: addToOnline,
                        removeFromOnline: removeFromOnline,
                        getOnline: getOnline,
                        getOnlineIds: getOnlineIds,
                        getOnlineExceptContact: getOnlineExceptContact,
                        sendMessage: sendMessage               
                    }

            function addToOnline(profile) {
                 
            } ;

            function removeFromOnline(user_id) {
                 alert('remove '+ user_id);
                 log($rootScope.online_user_list);
                    for (var i = 0; i < $rootScope.online_user_list.length; i++) {
                        if(user_id == $rootScope.online_user_list[i].user_id) {
                            $rootScope.online_user_list.splice(i,1);  
                        }
                        log($rootScope.online_user_list[i].user_id);
                    }                  
            } ;


            function setOnline() {
                 
            } ;

            function setOffline() {

            } ;

            function getOnlineExceptContact(callback) {
             
                var url = utils.prepare_url(apiconf.api.get_online_except_contact.url,{'[user_id]':$rootScope.currentUserId});
                return $http.get(url).success(callback); 

            };


            function getOnline(callback) {
             
                var url = utils.prepare_url(apiconf.api.get_online.url,{'[user_id]':$rootScope.currentUserId});
                return $http.get(url).success(callback); 

            };

            function getOnlineIds(callback) {
             
                var url = utils.prepare_url(apiconf.api.get_online_ids.url,{'[user_id]':$rootScope.currentUserId});
                return $http.get(url).success(callback); 

            };



            /*"""
            .. function:: sendMessage()

            Send message to inbox by http request to the server.
               
            */

            function sendMessage(owner_id,contact_id,message,callback) {
                             var url = utils.prepare_url(apiconf.api.send_message.url,{});
                             var data = {'app_name':local_config.app_name, 'owner_id':$rootScope.currentUserId,'contact_id':contact_id,  'message': message}
                             return $http.post(url,data).success(callback);
            };



    }]);


})();

File: djapp/chat/views_api/online.py
import json
from django.http import HttpResponse
from jsonview.decorators import json_view
from django.shortcuts import redirect
from utils.api_router import get_url_by_name
from django.views.decorators.csrf import csrf_exempt
import requests
from django.contrib.auth.models import User
from chat.models import ChatUser, ChatContacts
from chat.models import Tpa
from utils.util import serialize_user, get_url_by_name
from djapp.local import TPA_SERVER

@json_view
def get_online_except_contact(request,app_name,user_id):
    '''
    Function return list of users with opposite gender who are online except those who is not in contact list.

    [server]/api/[app_name]/[user_id]/get_online_except_contact

    Example: http://chat.localhost/api/tpa1com/150031/get_online_except_contact
    '''
    userlst_profile = []
    tpa = Tpa.objects.get(name=app_name)
    if user_id == 'undefined':
        #users_online = ChatUser.objects.filter(tpa=tpa,is_online=1)
        users_online = []
    else:
        owner = ChatUser.objects.get(tpa=tpa,user_id=user_id)
        users_online = ChatUser.objects.filter(tpa=tpa,is_online=1).exclude(gender=owner.gender)
        cc = ChatContacts.objects.filter(tpa=tpa,owner=owner)
        for c in cc:
            users_online = users_online.exclude(id=c.contact.id)
    for u in users_online:        
        userlst_profile.append(serialize_user(u))
    return { 'status': 0, 'message': 'ok', 'user_list': userlst_profile }




@json_view
def get_online(request,app_name,user_id):
    '''
    Function return list of users with opposite gender who are on line

    [server]/api/[app_name]/[user_id]/get_online

    Example: http://chat.localhost/api/tpa1com/150031/get_online
    '''
    userlst_profile = []
    tpa = Tpa.objects.get(name=app_name)
    if user_id == 'undefined':
        return { 'status': 0, 'message': 'ok', 'user_list': {} }
    else:
        try:
            owner = ChatUser.objects.get(tpa=tpa,user_id=user_id)
        except:
            url = get_url_by_name('get_profile_from_tpa',{'user_id':user_id,'app_name':app_name,'signal_server': TPA_SERVER})
            responce = requests.get(url)
            owner = ChatUser.objects.get(tpa=tpa,user_id=user_id)
        users_online = ChatUser.objects.filter(tpa=tpa,is_online=1).exclude(gender=owner.gender)
        for u in users_online:        
            userlst_profile.append(serialize_user(u))
    return { 'status': 0, 'message': 'ok', 'user_list': userlst_profile }



@json_view
def get_online_ids(request,app_name,user_id):
    '''
    Function return list of users (IDs) with opposite gender who are on line and not in contact list

    [server]/api/[app_name]/[user_id]/get_online_ids

    Example: http://chat.localhost/api/tpa1com/150031/get_online_ids
    '''
    userlst_profile = []
    tpa = Tpa.objects.get(name=app_name)
    owner = ChatUser.objects.get(tpa=tpa,user_id=user_id)
    users_online = ChatUser.objects.filter(tpa=tpa,is_online=1).exclude(gender=owner.gender)
    contacts = []
    for c in ChatContacts.objects.filter(owner=owner):
        contacts.append(int(c.contact.user_id))
    print contacts
    for u in users_online: 
        if not u.user_id in contacts:       
            userlst_profile.append(u.user_id)
    return { 'status': 0, 'message': 'ok', 'count': len(userlst_profile), 'user_list': userlst_profile }


@csrf_exempt
@json_view
def send_message(request):
    '''
    Function make a request to send message to the user's inbox'.

    parameters by POST: app_name,owner_id,contact_id,message

    [server]/api/send_message

    Example: http://chat.localhost/api/send_message
    '''
    #import pdb; pdb.set_trace() 
    b = json.loads(request.body)
    print b
    tpa = Tpa.objects.get(name=b['app_name'])
    owner = ChatUser.objects.get(tpa=tpa,user_id=int(b['owner_id']))

  
    return  { 'status': 0 }
    







